database_cleanup/models/purge_modules.py

```diff
@@ -50,9 +50,11 @@ def purge(self):
         if not modules:
             return True
         self.logger.info("Purging modules %s", ", ".join(module_names))
-        modules.filtered(
-            lambda x: x.state not in ("uninstallable", "uninstalled")
-        ).button_immediate_uninstall()
+        installed = modules.filtered(lambda x: x.state in ("installed", "to upgrade"))
+        ((modules - installed) + (modules - installed).downstream_dependencies()).write(
+            {"state": "to remove"}
+        )
+        installed.button_immediate_uninstall()
```

Comment:
this code is complex and not obvious about the intent.

If I read correctly, modules not in `installed`, `to upgrade` (filtered here) and  `uninstallable`, `installed`, `to remove` (default argument of downstream_dependencies) are excluded.
So it let: `uninstalled`  `to remove`, `to install`. Is it expected ? 

Plus, it looks like this code can be shorter if it use the right exclude_filter of  `modules.downstream_dependencies()`.

Reply:
@miikanissi Any comment on above question?

Reply:
I asked to add this in the previous PR: https://github.com/hbrunn/server-tools/commit/08c5df64c93de602ebef44d4e837e9e3b68bc2a5

That's because in v16 a test was added in this function that [raises](https://github.com/OCA/OCB/commit/0ee76617a7817e02fa845fc4c5e9e7663e3dc7ce) for modules that are not installed.

Looking at this more than a year later, I'd say don't call `button_immediate_uninstall` at all, write "to remove" on all of them and pass the empty set for the excluded states, as we want to purge stuff.

Calling `button_uninstall` is a relict from older versions where the actual cleanup happened there, by now that happens in `module_uninstall` called on `to remove` modules during the db reload we trigger below.